1199 - Add sync_batch_jobs Management Command to sync local job/dataset model states with AWS batch

nozomione opened the pull request:
## Issue Number

Closes #1199

## Purpose/Implementation Notes

I've implemented the `sync_batch_jobs` management command and added the corresponding test, `TestSyncBatchJobs`. 

Other changes include:
- Add new `JobStates` enum members:`SUCCEEDED` and `FAILED`
- In the `Dataset` model:
  - Added two new methods, used by the `Job` model:
     - ~~`apply_last_jobs`~~ `update_from_last_jobs` (class method): bulk updates dataset states based on the last jobs
     - ~~`on_uncaught_failure`~~ `on_failed` (instance method): marks a dataset as errored
- In the `Job` model:
  - Replaced the  `update_job_state` method with two new methods:
    - `get_job_state`(static method): maps a local job state to its corresponding AWS Batch job status
    - ~~`apply_state_at`~~ `update_state_at` (instance method): updates timestamp fields,(`*_at`) based on a new synced job state
  - Updated the following existing methods and their corresponding tests to reflect these changes:
    -  `terminate` 
    - `terminate_submitted` 
    -  `sync_state`
    - `bulk_sync_state`

> [!note] 
> Please see the following comments for additional changes included in this PR:
> - [Comment 1](https://github.com/AlexsLemonade/scpca-portal/pull/1208#issuecomment-2825588048)
> - [Comment 2](https://github.com/AlexsLemonade/scpca-portal/pull/1208#issuecomment-2837141925)

## Types of changes

- Refactor (addresses code organization and design mentioned in corresponding issue)
- New feature (non-breaking change which adds functionality)

## Functional tests

Updated:
- `TestJob::test_sync_state`
- `TestJob::test_bulk_sync_state`
- `TestJob::test_bulk_sync_state_no_matching_batch_job_found`
- `TestJob::assertDatasetState` (a new helper method for `job.dataset` assertions)

Newly Added:
- `TestSyncBatchJobs`

## Checklist

<!-- Put an `x` in the boxes that apply. -->

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

nozomione commented:
We'll introduce the following changes:
- Replace `JobStates.COMPLETED` with `JobStates.SUCCEEDED` and `JobStates.FAILED` to align with AWS Batch statuses
- Remove the `Job.terminated_at` field and use `Job.completed_at` to record when jobs are terminated
- Replace the static`Job::update_job_state` method with the following new methods:
   - A static method `Job::get_jobs_state`: Maps a local job state to its corresponding AWS Batch job status
   - An instance method `Job::apply_state_at`: Updates timestamp fields, `*_at`, based on a new synced job state
- Add a new class method `Job::bulk_sync_state`:  Updates newly synced jobs in the database in bulk.
- Add a new class method `Dataset::apply_last_jobs`: Updates the fields of associated datasets' based on new synced job states

nozomione commented:
We'll introduce the following changes:
- Replace `JobStates.COMPLETED` with `JobStates.SUCCEEDED` and `JobStates.FAILED` to align with AWS Batch statuses
- Remove the `Job.terminated_at` field and use `Job.completed_at` to record when jobs are terminated
- Replace the static`Job::update_job_state` method with the following new methods:
   - `Job::get_jobs_state`: A static method which maps a local job state to its corresponding AWS Batch job status
   - `Job::apply_state_at`: An instance method which updates timestamp fields, `*_at`, based on a new synced job state
- Add a new `Job::bulk_sync_state` class method which updates newly synced jobs in the database in bulk.
- Add a new `Dataset::apply_last_jobs` class method which updates the fields of associated datasets' based on new synced job states

nozomione commented on api/scpca_portal/test/management/commands/test_sync_batch_jobs.py:
```diff
@@ -0,0 +1,104 @@
+            error_message="Job FAILED",
+        )
+
+        # TERMINATED job state and dataset should be updated
+        terminated_job = Job.objects.filter(batch_job_id="MOCK_JOB_ID_007").first()
+        self.assertEqual(terminated_job.state, JobStates.TERMINATED)
+        self.assertIsNotNone(failed_job.failure_reason)
+        self.assertDatasetState(
+            terminated_job.dataset,
+            is_processing=False,
+            is_errored=False,
+            errored_at=None,
+            error_message=None,
+        )
```
**Q:** We assign the `Job`s' `failure_reason` fields to both `FAILED` and `TERMINATED` jobs, but we only set `dataset.error_message` for `FAILED` jobs. Should we rename `failure_reason` to `status_reason` for better clarity?

nozomione replied:
**Q:** We assign the `Job`s' `failure_reason` field to both `FAILED` and `TERMINATED` jobs, but we only set `dataset.error_message` for `FAILED` jobs. Should we rename `failure_reason` to `status_reason` for better clarity?

davidsmejia commented on api/scpca_portal/models/dataset.py:
```diff
@@ -91,6 +94,25 @@ class Meta:
     def __str__(self):
         return f"Dataset {self.id}"
 
+    @classmethod
+    def apply_last_jobs(cls, synced_jobs: Iterable):
+        """
+        Bulk updates the states based on the given latest synced jobs.
+        Each dataset is marked as errored if its corresponding job has failed.
+        """
+        synced_datasets = []
+
+        for job in synced_jobs:
+            if job.state == JobStates.FAILED:
+                job.dataset.on_uncaught_failure(job.failure_reason)
+
+            job.dataset.is_processing = False
+            synced_datasets.append(job.dataset)
+
+        cls.objects.bulk_update(
+            synced_datasets, ["is_processing", "is_errored", "errored_at", "error_message"]
+        )
+
```
Can we update this so that we are handling any arbitrary collection of jobs.

davidsmejia commented on api/scpca_portal/models/job.py:
```diff
@@ -224,46 +212,76 @@ def terminate_submitted(cls) -> List[Self]:
+            ["state", "failure_reason", "completed_at"],
+        )
+
+        Dataset.apply_last_jobs(synced_jobs)
```
Pass datasets here.

nozomione replied:
I've left it as is and created a mapping object to retrieve the corresponding jobs from `update_from_last_jobs`. Let me know your thoughts on this!

davidsmejia commented on api/scpca_portal/enums/job_states.py:
```diff
@@ -4,5 +4,7 @@
 class JobStates(TextChoices):
```
We should update to have one corresponding timestamp for each state.

If we update an "upstream" timestamp we should wipe out the "downstream" timestamps. ie. if we update `succeeded_at` we should nullify `terminated_at` because it no longer applies to that dataset.

nozomione replied:
I've added a new static method, `Dataset::reset_state`, to reset the following states to their default values:
- `is_processing` 
- `is_error` (could be retried depending on the error?)
- `is_terminated`

nozomione replied:
I've added a new instance method, `Dataset::reset_state`, to reset the following states to their default values:
- `is_processing` 
- `is_error` (could be retried depending on the error?)
- `is_terminated`  

Let me know your thoughts on this!

nozomione replied:
I've added a new instance method, `Dataset::reset_state`, which resets the following fields to their default values for retrying jobs on a dataset instance:

- `is_processing`
- `is_error`, `error_message` (can be retried depending on the error unless irrecoverable?)
- `is_terminated`

The `Dataset::apply_state_at` method handles resetting the timestamps before reapplying new ones based on the last job.

I've excluded `is_processed` and `processed_at`, since a successfully processed dataset won't be retried, instead, it will be regenerated with a new dataset ID if it expires (though this behavior may change?).

Let me know what you think! Thank you, David!

nozomione replied:
**UPDATE:** The `Dataset` states and timestamps fields will be reset and overwritten based on the latest job, regardless of whether the previous job completed successfully. Any datasets, including pre-generated `CCDL` or user-generated ones, can be reprocessed with a new job using the same dataset ID. 

(NOTE: Implementation details are subject to change.)

nozomione replied:
> we change the state to be "PROCESSING".

Great 👍

nozomione commented:
I've made the following changes:

- In  the `Dataset` model:
  - Added new fields for `JobStates.TERMINATED`:
    - `is_terminated`
    - `terminated_at`
  - Added new methods:
    - `update_from_last_job` (instance) - called from the `update_from_last_jobs` class method (previously `apply_last_jobs`)
    - `on_succeeded`, `on_failed` (previously `on_uncaught_failure`), `on_terminated` (instance) - update dataset state
    - `apply_state_at` (instance) - updates `*_at` timestamps fields
    - `reset_state` (static) - rests state fields to their default values

- In  the `Job` model:
  - Added new fields, `succeeded_at` and `failed_at` to align with `JobStates.SUCCEEDED` and  `JobStates.FAILED`, respectively 
  - Updated methods:
    - `bulk_update_state`
    - `bulk_sync_state`, `sync_state`
    - `apply_state_at`
  
- Adjusted tests:
  - `TestJob`
  - `TestSyncBatchJobs` 

This PR is ready for another look. Thank you, David!

nozomione commented:
I've applied your feedback and made the following changes:

- In  the `Dataset` model:
  - Added new fields for `JobStates.TERMINATED`:
    - `is_terminated`
    - `terminated_at`
  - Added new methods:
    - `update_from_last_job` (instance) - called from the `update_from_last_jobs` class method (previously `apply_last_jobs`)
    - `on_succeeded`, `on_failed` (previously `on_uncaught_failure`), `on_terminated` (instance) - update dataset state
    - `apply_state_at` (instance) - updates `*_at` timestamps fields
    - `reset_state` (static) - rests state fields to their default values

- In  the `Job` model:
  - Added new fields, `succeeded_at` and `failed_at` to align with `JobStates.SUCCEEDED` and  `JobStates.FAILED`, respectively 
  - Updated the following methods:
    - `bulk_update_state`
    - `bulk_sync_state`, `sync_state`
    - `apply_state_at`
  
- Adjusted the following tests:
  - `TestJob`
  - `TestSyncBatchJobs` 

This PR is ready for another look. Thank you, David!

nozomione commented:
I've updated the PR (see https://github.com/AlexsLemonade/scpca-portal/pull/1208#issuecomment-2837141925), and this is ready for another look. Thank you, David!

avrohomgottlieb commented on api/scpca_portal/models/job.py:
```diff
@@ -224,46 +215,90 @@ def terminate_submitted(cls) -> List[Self]:
+                    job.failure_reason = failure_reason
+                    job.update_state_at()
+                    synced_jobs.append(job)
+                else:
+                    continue
```
```suggestion

```

avrohomgottlieb commented on api/scpca_portal/models/job.py:
```diff
@@ -224,46 +215,90 @@ def terminate_submitted(cls) -> List[Self]:
+                else:
+                    continue
+
+        if synced_jobs:
+            cls.bulk_update_state(synced_jobs)
+            return True
 
         return False
```
I think this fits better with the pattern throughout the method. Also, it's typically good to follow the happy path.

```suggestion
        if not synced_jobs:
            return False

        cls.bulk_update_state(synced_jobs)
        return True
```

nozomione replied:
I’ve updated the code to follow that pattern - thank you!

nozomione replied:
I’ve updated the code to follow that happy path pattern in both `bulk_update_state`and `sync_state` - thank you!

avrohomgottlieb commented on api/scpca_portal/models/dataset.py:
```diff
@@ -108,6 +113,32 @@ def get_or_find_ccdl_dataset(
+        """
+
+        for dataset in datasets:
+            dataset.update_from_last_job(save=not bulk_save)
```
```suggestion
            dataset.update_from_last_job(save(=not bulk_save))
```

nozomione replied:
We've removed `bulk_save` and now explicitly pass `False` to this call - thank you for the helpful tip!

avrohomgottlieb commented on api/scpca_portal/models/dataset.py:
```diff
@@ -108,6 +113,32 @@ def get_or_find_ccdl_dataset(
+        for dataset in datasets:
+            dataset.update_from_last_job(save=not bulk_save)
+
+        if bulk_save:
+            cls.objects.bulk_update(
+                datasets,
+                [
+                    "is_processing",
+                    "is_processed",
+                    "processed_at",
+                    "is_errored",
+                    "errored_at",
+                    "error_message",
+                    "is_terminated",
+                    "terminated_at",
+                ],
+            )
```
```suggestion
        if bulk_save:
              updated_attrs = [
                      "is_processing",
                      "is_processed",
                      "processed_at",
                      "is_errored",
                      "errored_at",
                      "error_message",
                      "is_terminated",
                      "terminated_at",
                ]
                cls.objects.bulk_update(datasets, updated_attrs)
            )
```

avrohomgottlieb reviewed (commented):
This is great, Nozomi. Left a few small pieces of feedback, but I think you did a great job.

nozomione commented:
I've updated the codebase based on the feedback (also edited the [comment](https://github.com/AlexsLemonade/scpca-portal/pull/1208#issuecomment-2837141925) to reflect the new additional changes we discussed today). 

This PR is ready for another look. Thank you David & Avrohom!
